pkg/downloader/chart_downloader.go

```diff
@@ -158,6 +159,66 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 	return destfile, ver, nil
 }
 
+type inMemoryChart struct {
+	ChartData              *bytes.Buffer
+	Version                string
+	ProvenanceVerification *provenance.Verification
+}
+
+type InMemoryChartList map[string]inMemoryChart
+
+var inMemoryCache = make(InMemoryChartList)
+
+//DownloadToInMemoryCache takes the chart reference and version and downloads it to the inMemoryCache for usage.
+//
+//The chart downloaded to the inMemoryCache can either be from the filesystem or from a URL
+func (c *ChartDownloader) DownloadToInMemoryCache(ref, version string) (string, *provenance.Verification, error) {
+	u, err := c.ResolveChartVersion(ref, version)
+	if err != nil {
+		return "", nil, err
+	}
+
+	g, err := c.Getters.ByScheme(u.Scheme)
+	if err != nil {
+		return "", nil, err
+	}
+
+	chartData, err := g.Get(u.String(), c.Options...)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// If provenance is requested, verify it.
+	ver := &provenance.Verification{}
+	if c.Verify > VerifyNever {
+		provData, err := g.Get(u.String() + ".prov")
+		if err != nil {
+			if c.Verify == VerifyAlways {
+				return ref, ver, errors.Errorf("failed to fetch provenance %q", u.String()+".prov")
+			}
+			fmt.Fprintf(c.Out, "WARNING: Verification not found for %s: %s\n", ref, err)
+			return ref, ver, nil
+		}
+
+		if c.Verify != VerifyLater {
+			ver, err = verifyInMemoryChart(ref, c.Keyring, chartData, provData)
+			if err != nil {
+				// Fail always in this case, since it means the verification step
+				// failed.
+				return ref, ver, err
+			}
+		}
+	}
+
+	inMemoryCache[ref] = inMemoryChart{
```

Comment:
Getting back to the concern that `ref` ignores the chart version.

I'm under the impression that when a chart is downloaded, the version is taken into account.  See here
https://github.com/helm/helm/blob/master/pkg/downloader/chart_downloader.go#L257

If this is true, then we need to know if multiple versions of the same chart can be used as dependencies.  If so, the index of the cache must consider the version also.